components/table/table-wrapper.js

```diff
@@ -328,14 +328,33 @@ export class TableWrapper extends RtlMixin(LitElement) {
 
 		// observes mutations to <table>'s direct children and also
 		// its subtree (rows or cells added/removed to any descendant)
-		this._tableObserver = new MutationObserver(() => this._applyClassNames(table));
-		this._tableObserver.observe(table, {
+		this._tableMutationObserver = new MutationObserver(() => this._applyClassNames(table, 0));
+		this._tableMutationObserver.observe(table, {
 			attributes: true, /* required for legacy-Edge, otherwise attributeFilter throws a syntax error */
 			attributeFilter: ['selected'],
 			childList: true,
 			subtree: true
 		});
-		this._applyClassNames(table);
+
+		// observes when visibility of <table> changes to catch cases
+		// where table is initially hidden (which would cause first/last
+		// row classes to be unset) but then becomes shown
+		if (typeof(IntersectionObserver) === 'function') {
+			if (this._tableIntersectionObserver === null) {
+				this._tableIntersectionObserver = new IntersectionObserver((entries) => {
+					entries.forEach((entry) => {
+						if (entry.isIntersecting) {
+							this._tableIntersectionObserver.unobserve(table);
+							requestIdleCallback(() => this._applyClassNames(table));
+						}
+					});
+				});
+			} else {
+				this._tableIntersectionObserver.disconnect();
+			}
+			this._tableIntersectionObserver.observe(table);
+		}
+		requestAnimationFrame(() => this._applyClassNames(table));
```

Comment:
Is this rAF in case `IntersectionObserver` is not supported?  I did a quick test and it seems that the observer gets called initially, not just after page-load. I seem to recall some flakiness with Safari when using it to detect `display` changes a year or two ago... unable to reproduce now though so maybe they fixed some stuff.

Reply:
I originally only had this in the `else` branch -- so it would only run if `IntersectionObserver` wasn't supported. But that lead to it not running right away if a table was outside the viewport... which initially seems OK but was causing most of the visual diff tests to fail (since they're long pages and not in view) but also some initially jitter when you scroll them into view.